mitmproxy/proxy/protocol2/websocket.py

```diff
@@ -16,14 +19,84 @@ def __init__(self, context: ClientServerContext, handshake_flow: http.HTTPFlow):
+
+        self.client_conn = wsconn.WSConnection(wsconn.SERVER, wsconn.ConnectionState.OPEN, extensions=[extension])
+        self.server_conn = wsconn.WSConnection(wsconn.CLIENT, wsconn.ConnectionState.OPEN, extensions=[extension])
 
     @expect(events.Start)
     def start(self, _) -> commands.TCommandGenerator:
-        yield from ()
+        yield commands.Hook("websocket_start", self.flow)
         self._handle_event = self.relay_messages
 
+    _handle_event = start
+
     @expect(events.DataReceived, events.ConnectionClosed)
     def relay_messages(self, event: events.Event) -> commands.TCommandGenerator:
-        raise NotImplementedError()
+        if isinstance(event, events.DataReceived):
+            from_client = event.connection == self.context.client
+            if from_client:
+                source = self.client_conn
+                other = self.server_conn
+                fb = self.client_frame_buffer
+                send_to = self.context.server
+            else:
+                source = self.server_conn
+                other = self.client_conn
+                fb = self.server_frame_buffer
+                send_to = self.context.client
 
-    _handle_event = start
+            source.receive_bytes(event.data)
+
+            for ws_event in source.events():
+                if isinstance(ws_event, wsevents.DataReceived):
+                    fb.append(ws_event.data)
+                    if ws_event.message_finished:
+                        payload = b''.join(f for f in fb)
+                        fb.clear()
+                        websocket_message = websocket.WebSocketMessage(0x1 if isinstance(ws_event, wsevents.TextReceived) else 0x2,
+                                                                       from_client, payload)
+                        self.flow.messages.append(websocket_message)
+                        yield commands.Hook("websocket_message", self.flow)
+
+                    other.send_data(ws_event.data, ws_event.message_finished)
+                    yield commands.SendData(send_to, other.bytes_to_send())
+
+                elif isinstance(ws_event, wsevents.ConnectionClosed):
+                    yield commands.SendData(send_to, source.bytes_to_send())
+                    if isinstance(ws_event, wsevents.ConnectionFailed):
+                        self.flow.error = flow.Error(
+                            "WebSocket connection closed unexpectedly by {}: {}".format("client" if from_client else "server",
+                                                                                        ws_event.reason)
+                        )
+                        yield commands.Hook("websocket_error", self.flow)
+
+                    yield commands.Hook("websocket_end", self.flow)
+                    self._handle_event = self.done
+        elif isinstance(event, events.ConnectionClosed):
+            if event.connection == self.context.server:
```

Comment:
Is this an expected event according to protocol? My understanding would be that receiving `ConnectionClosed` in `relay_messages` means ungraceful shutdown of one end, as a graceful shutdown would first cause a `wsevents.ConnectionClosed`? In this case, we should add an error here.